p2p/src/swarm/peerdb.rs

```diff
@@ -30,86 +30,119 @@
+// TODO: store available peers into a binary heap
 pub struct PeerDb<T: NetworkingService> {
-    /// Hashmap for peer information
-    peers: HashMap<T::PeerId, PeerContext<T>>,
+    /// P2P configuration
+    p2p_config: Arc<config::P2pConfig>,
 
-    /// Hashmap of discovered peers we don't have an active connection with
-    discovered: HashMap<T::PeerId, PeerAddrInfo<T>>,
+    /// Set of peers known to `PeerDb`
+    peers: HashMap<T::PeerId, Peer<T>>,
+
+    /// Set of available (idle) peers
+    available: HashSet<T::PeerId>,
 
     /// Pending connections
-    pending: HashMap<T::Address, PeerAddrInfo<T>>,
+    pending: HashMap<T::Address, T::PeerId>,
 
     /// Banned peers
     banned: HashSet<T::PeerId>,
 }
 
 impl<T: NetworkingService> PeerDb<T> {
-    pub fn new() -> Self {
+    pub fn new(p2p_config: Arc<config::P2pConfig>) -> Self {
         Self {
-            peers: HashMap::new(),
-            discovered: HashMap::new(),
-            pending: HashMap::new(),
-            banned: HashSet::new(),
+            peers: Default::default(),
+            available: Default::default(),
+            pending: Default::default(),
+            banned: Default::default(),
+            p2p_config,
         }
     }
 
     /// Get the number of idle (available) peers
     pub fn idle_peer_count(&self) -> usize {
-        self.discovered.len()
+        self.available.len()
     }
 
-    /// Get socket address of the next best peer (TODO: in terms of peer score)
-    pub fn best_peer_addr(&mut self) -> Option<T::Address> {
-        // TODO: improve peer selection
-        let key = match self.discovered.keys().next() {
-            Some(key) => *key,
-            None => return None,
-        };
+    /// Get the number of active peers
+    pub fn active_peer_count(&self) -> usize {
+        self.peers
+            .len()
+            .saturating_sub(self.pending.len())
+            .saturating_sub(self.available.len())
+            .saturating_sub(self.banned.len())
+    }
 
-        // TODO: find a better way to store the addresses
-        let peer_info = self.discovered.remove(&key).expect("peer to exist");
-        let addr = match peer_info {
-            PeerAddrInfo::Raw { ref ip4, ref ip6 } => {
-                assert!(!(ip4.is_empty() && ip6.is_empty()));
+    pub fn active_peers(&self) -> Vec<(&T::PeerId, &PeerContext<T>)> {
+        self.peers
+            .iter()
+            .filter_map(|(id, info)| match info {
+                Peer::Active(inner) => Some((id, inner)),
+                _ => None,
+            })
+            .collect::<Vec<_>>()
+    }
 
-                if ip6.is_empty() {
-                    ip4.iter().next().expect("ip4 empty").clone()
-                } else {
-                    ip6.iter().next().expect("ip6 empty").clone()
-                }
-            }
-        };
+    /// Get mutable reference to the peer store
+    pub fn peers(&mut self) -> &mut HashMap<T::PeerId, Peer<T>> {
```

Comment:
Why do all these functions have to be mutable like this? This is quite dangerous... the struct should manage what access rules are allowed.

Reply:
Agreed with @TheQuantumPhysicist. If you really want a mutable reference, I'd prefer the method to be called `peers_mut` so it's more more clear what's going on in use sites.

Reply:
I can probably get rid of these

Reply:
Even `peers_mut` is so invasive. IMHO, I think the right way is to get use getters/setters that limit that possible changes to this struct so that invariants can remain correct.